Add timeouts and an enable/disable switch

jordan-brough opened the pull request:
PR for https://github.com/solidusio/solidus_avatax/issues/15.

cc @magnusvk

jordan-brough commented:
Note: Having the enable/disable switch inside of solidus_avatax in this way feels a bit weird to me, but I'm OK with it if others are.  I could also see us just trying to make it really easy to add your own enable/disable logic inside of the applications themselves if people thought that was better.

athal7 commented on app/models/spree_avatax/shared.rb:
```diff
@@ -28,6 +28,24 @@ def taxable_order?(order)
+      end
+    end
+
     def tax_svc
```
can this become private?

jordan-brough replied:
Yah it could but I was thinking it's probably nice to be able to access this from a console & etc, or from custom code if someone using Avatax is using an endpoint that we're not covering.

athal7 commented on lib/tasks/commit_backfill.rake:
```diff
@@ -63,7 +56,7 @@ namespace :spree_avatax do
       }
 
-      history_result = tax_svc.gettaxhistory(history_request)
+      history_result = SpreeAvatax::Shared.tax_svc.gettaxhistory(history_request)
```
these might need to change if we make that tax_svc private per the suggestion above

jordan-brough replied:
Yeah, I think it would be good to actually keep `tax_svc` public as an easy way for client to hit Avatax endpoints we haven't used/tested/etc.

athal7 commented:
looking good to me! and nice cleanups along the way.

magnusvk commented on app/models/spree_avatax/config.rb:
```diff
@@ -11,6 +13,17 @@ class << self
       attr_accessor :sales_invoice_cancel_error_handler
+
+      def timeout
+        (config = last) ? config.timeout : DEFAULT_TIMEOUT
```
It makes me just a tad uncomfortable to use `last` without an `order`. I know that this will order by id by default, but personally I'd love to be more explicit about this. Not a blocker, just a thought.

jordan-brough replied:
Sure, I'm fine with adding a `most_recent` helper method or something like that.

magnusvk commented on app/models/spree_avatax/shared.rb:
```diff
@@ -37,6 +55,12 @@ def tax_svc
     end
 
+    def with_timeout
+      Timeout.timeout(SpreeAvatax::Config.timeout, SpreeAvatax::AvataxTimeout) do
```
We should mention that we looked at the Avatax gem to make sure they don't use any sort of connection pooling or the like which might cause problems with the use of `Timeout.timeout`.

jordan-brough replied:
Sounds good.

magnusvk commented:
Nice work, thank you. :+1:

jordan-brough commented:
@cbrunsdon or @jhawthorn or @gmacdougall any thoughts on the general approach here before I start doing tweaks & cleanups?

jordan-brough commented:
@magnusvk @athal7 I implemented some of those suggestions and refactored the timeout code a little bit.  Good to go?  Stembolt says they are fine with us managing solidus_avatax.

magnusvk commented:
:+1: :ship: :it:
